nearcore/src/runtime/mod.rs

```diff
@@ -1078,26 +1078,29 @@ impl RuntimeAdapter for NightshadeRuntime {
         next_epoch_shard_layout: &ShardLayout,
         state_changes: StateChangesForSplitStates,
     ) -> Result<Vec<ApplySplitStateResult>, Error> {
-        let trie_changes = self.tries.apply_state_changes_to_split_states(
+        let trie_updates = self.tries.apply_state_changes_to_split_states(
             &state_roots,
             state_changes,
             &|account_id| account_id_to_shard_uid(account_id, next_epoch_shard_layout),
         )?;
 
-        Ok(trie_changes
-            .into_iter()
-            .map(|(shard_uid, trie_changes)| ApplySplitStateResult {
+        let mut applied_split_state_results: Vec<_> = vec![];
+        for (shard_uid, trie_update) in trie_updates {
+            let (_, trie_changes, state_changes) = trie_update.finalize()?;
+            applied_split_state_results.push(ApplySplitStateResult {
                 shard_uid,
                 new_root: trie_changes.new_root,
                 trie_changes: WrappedTrieChanges::new(
                     self.get_tries(),
                     shard_uid,
                     trie_changes,
-                    vec![],
+                    state_changes,
```

Comment:
Do we have a strong guarantee that all state_changes here will be marked as Resharding? Is there any sensible way to enforce that condition (static(types), dynamic(asserts), comments in code or otherwise)?

Reply:
The original code had a debug assert so it makes sense to have a debug assert here as well to check if state_changes are always committed with cause as resharding. Don't really want to have a dynamic error here as we need to ensure in code that the commit cause is resharding. Let me know if you feel differently.

Although that said, I don't think the performance cost would be too much to convert this from a debug assert.